src/ProductUsageService.php

```diff
@@ -0,0 +1,251 @@
+<?php
+
+namespace Drupal\amazon_product_widget;
+
+use Drupal\amazon_product_widget\Plugin\Field\FieldType\AmazonProductField;
+use Drupal\Core\Database\Connection;
+use Drupal\Core\Entity\EntityFieldManagerInterface;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Entity\FieldableEntityInterface;
+use Drupal\Core\Extension\ModuleHandlerInterface;
+
+/**
+ * Allows for adding or deleting product usages.
+ *
+ * @package Drupal\amazon_product_widget
+ */
+class ProductUsageService {
+
+  /**
+   * Database connection.
+   *
+   * @var \Drupal\Core\Database\Connection
+   */
+  protected $database;
+
+  /**
+   * EntityFieldManager.
+   *
+   * @var \Drupal\Core\Entity\EntityFieldManagerInterface
+   */
+  protected $fieldManager;
+
+  /**
+   * Module handler.
+   *
+   * @var \Drupal\Core\Extension\ModuleHandlerInterface
+   */
+  protected $moduleHandler;
+
+  /**
+   * ProductUsageService constructor.
+   *
+   * @param \Drupal\Core\Database\Connection $database
+   *   Database connection.
+   * @param \Drupal\Core\Entity\EntityFieldManagerInterface $fieldManager
+   *   EntityFieldManager.
+   * @param \Drupal\Core\Extension\ModuleHandlerInterface $moduleHandler
+   *   Module handler.
+   */
+  public function __construct(Connection $database, EntityFieldManagerInterface $fieldManager, ModuleHandlerInterface $moduleHandler) {
+    $this->database = $database;
+    $this->fieldManager = $fieldManager;
+    $this->moduleHandler = $moduleHandler;
+  }
+
+  /**
+   * Updates ASIN mapping for the given entity.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   Entity to update the mapping for.
+   */
+  public function update(EntityInterface $entity) {
+    if (!$entity instanceof FieldableEntityInterface) {
+      return;
+    }
+
+    if ($entity->getEntityTypeId() === 'paragraph') {
+      return;
+    }
+
+    // MenuLinkEntity returns NULL on call to $entity->id() so we check here to
+    // be sure.
+    if ($entity->id() === NULL) {
+      return;
+    }
+
+    $this->purge($entity->id(), $entity->getEntityTypeId());
```

Comment:
so this will fragment the database table, since this is run frequently we should add a cron job, eg. daily, or maybe weekly to optimize table

Reply:
also before purging, first we should check whether an update (purge/install) is even necessary, ie. any asins have changed in the first place